include/base/ghosting_functor.h

```diff
@@ -157,13 +158,47 @@ class GhostingFunctor : public ReferenceCountedObject<GhostingFunctor>
   /**
    * Constructor.  Empty in the base class.
    */
-  GhostingFunctor() {}
+  GhostingFunctor(): _mesh(nullptr) {}
+
+  /**
+   * Constructor using mesh
+   */
+  GhostingFunctor(const MeshBase & mesh): _mesh(&mesh) {}
+
+  /**
+   * Copy Constructor
+   */
+  GhostingFunctor(const GhostingFunctor & other) :
+   ReferenceCountedObject<GhostingFunctor>(other),
+   _mesh (other._mesh)
```

Comment:
This was the "bug" that incited this whole PR right? Would it even make sense to clone the same ghosting functor onto the same mesh? would it be safer to leave `_mesh` as `nullptr` and then put in protective asserts in places where `_mesh` is used? (If they're not already there, they should be anyways right?)

Reply:
I get that in the code you've added you immediately call `set_mesh` after `clone`, but maybe someone else will use this in a different way in the future. Unlikely, but I'm always paranoid

Reply:
Great point!!! 

If it is not too evil to assume that a ghosting functor highly likely has  a mesh, then it might not be a bad idea to have clone function like:

`std::unique_ptr<GhostingFunctor> clone (const MeshBase * mesh) `

But this will contradict the definition of `clone`?? Possibly not.

```
  GhostingFunctor(const GhostingFunctor & other, const MeshBase * mesh) :
   ReferenceCountedObject<GhostingFunctor>(other),
   _mesh (mesh)
```

Reply:
> Would it even make sense to copy/clone the same ghosting functor onto the same mesh?

As a use case, usually no, though I could see a case where the same general functor might be copied and then the two copies given different (but neither superceding the other) settings.

As a general principle, though, I think it's the only thing that makes sense - when you make a copy of an object you want it to be as indistinguishable from the original as possible until/unless you start changing things afterwards.

> put in protective asserts in places where _mesh is used?

This, on the other hand, is such a good idea that I feel stupid for not having done so already.  I'd love it if, for any ghosting functor subclass that uses ```_mesh``` (and maybe even the ones that don't!), whenever the functor receives a non-empty range, it does something like ```libmesh_assert(_mesh.query_elem_ptr((*it)->id()) == *it);``` on each element ```*it``` in the range.  If we'd had that to begin with then it wouldn't have taken so long to track this bug down.

Reply:
Yes, I might add these changes